add events-executor and timers-manager in rclcpp

alsora opened the pull request:
This PR introduces the events-executor based on the design here https://github.com/ros2/design/pull/305 and the already-existing open-source implementation in https://github.com/irobot-ros/events-executor.


FYI @clalancette @wjwwood @mjcarroll @mauropasse @bpwilcox 


This PR currently targets Michael's branch as it depends on it. 
Will target the main branch after this PR is merged. 
https://github.com/ros2/rclcpp/pull/2143

alsora commented on rclcpp/test/rclcpp/executors/test_events_executor.cpp:
I think that most (all?) of the tests here can probably moved to the `test_executors.cpp` file as they are valid tests for all executors.

mauropasse commented on rclcpp/include/rclcpp/experimental/executors/events_executor/events_executor.hpp:
```diff
@@ -0,0 +1,284 @@
+/**
+ * This executor uses an events queue and a timers manager to execute entities from its
+ * associated nodes and callback groups.
+ * The RMW listener APIs are used to collect new events.
```
Also events can be generated from timers, guard conditions, (waitables?) anything which accepts a "on_ready" callback basically.

alsora replied:
Done in [141ea25](https://github.com/ros2/rclcpp/pull/2155/commits/141ea258dc8cab3b716aa5aa55b06b3cb049fd10)

mauropasse commented on rclcpp/include/rclcpp/experimental/executors/events_executor/events_executor_event_types.hpp:
```diff
@@ -0,0 +1,46 @@
+struct ExecutorEvent
+{
+  const void * exec_entity_id;
+  int gen_entity_id;
```
Not sure if `gen_entity_id` is a great name for this, could we call it `sub_entity_id`?
Or add a comment like `gen_entity_id` is the ID of the sub-entity generating the event. This is used on waitables having different entities which can generate an event, like .. "

alsora replied:
Done in [141ea25](https://github.com/ros2/rclcpp/pull/2155/commits/141ea258dc8cab3b716aa5aa55b06b3cb049fd10)

alsora commented on rclcpp/include/rclcpp/experimental/executors/events_executor/events_executor.hpp:
```diff
@@ -0,0 +1,284 @@
+  std::shared_ptr<rclcpp::executors::ExecutorNotifyWaitable> notify_waitable_;
+
+  /// Flag used to reduce the number of unnecessary waitable events
+  std::atomic<bool> notify_waitable_event_pushed_ {false};
```
This is used in `events_executor.cpp`

mauropasse commented on rclcpp/include/rclcpp/experimental/executors/events_executor/simple_events_queue.hpp:
```diff
@@ -0,0 +1,134 @@
+private:
+  // The underlying queue implementation
+  std::queue<rclcpp::experimental::executors::ExecutorEvent> event_queue_;
+  // Mutex to protect the insertion/extraction of events in the queue
```
Is not just insertion/extraction, is also checking size and if empty:
```suggestion
  // Mutex to protect concurrent actions over the queue
```

mauropasse commented on rclcpp/include/rclcpp/experimental/timers_manager.hpp:
```diff
@@ -0,0 +1,539 @@
+   *
+   * @param context custom context to be used.
+   * Shared ownership of the context is held until destruction.
+   * @param on_ready_callback The timers on ready callback. if not callable,
```
```suggestion
   * @param on_ready_callback The timers on ready callback. If not callable,
```

mauropasse replied:
Shall we add a note about the `on_ready_callback` and the difference from the callback called when a timer is executed? Like `Note: This callback is not the one called when timer is executed. It can be used to notify that a timer has to be executed.`

alsora replied:
Done in [141ea25](https://github.com/ros2/rclcpp/pull/2155/commits/141ea258dc8cab3b716aa5aa55b06b3cb049fd10)

mauropasse commented on rclcpp/include/rclcpp/experimental/timers_manager.hpp:
```diff
@@ -0,0 +1,539 @@
+
+      return removed;
+    }
+
```
```suggestion
    /**
     * @brief Retrieve the timer corresponding to an ID
     *
     * @param timer_id The ID of the timer to retrieve.
     * @return TimerPtr if there's a timer associated with the ID, nullptr otherwise
     */
```

alsora replied:
Done in [141ea25](https://github.com/ros2/rclcpp/pull/2155/commits/141ea258dc8cab3b716aa5aa55b06b3cb049fd10)

mauropasse commented on rclcpp/include/rclcpp/experimental/timers_manager.hpp:
```diff
@@ -0,0 +1,539 @@
+      std::make_heap(owned_heap_.begin(), owned_heap_.end(), timer_greater);
+    }
+
+    void clear_callbacks()
```
This API lacks doc. But also the name could be misleading or needs more info, since "callbacks" could refer to `on_ready` or `on_reset` callbacks. In this case, we clear here the `on_reset` ones.

alsora replied:
Done in [141ea25](https://github.com/ros2/rclcpp/pull/2155/commits/141ea258dc8cab3b716aa5aa55b06b3cb049fd10)

mauropasse commented on rclcpp/include/rclcpp/experimental/timers_manager.hpp:
```diff
@@ -0,0 +1,539 @@
+     */
+    static bool timer_greater(TimerPtr a, TimerPtr b)
+    {
+      // FIXME!
```
What was to be fixed here?

alsora replied:
This would result in an error if timers have a different clock associated with them.
It should be possible to enforce this when timers are added

alsora replied:
Done in [141ea25](https://github.com/ros2/rclcpp/pull/2155/commits/141ea258dc8cab3b716aa5aa55b06b3cb049fd10)

mauropasse commented on rclcpp/include/rclcpp/experimental/executors/events_executor/events_executor.hpp:
```diff
@@ -0,0 +1,284 @@
+  std::shared_ptr<rclcpp::executors::ExecutorNotifyWaitable> notify_waitable_;
+
+  /// Flag used to reduce the number of unnecessary waitable events
+  std::atomic<bool> notify_waitable_event_pushed_ {false};
```
I deleted the comment, `notify_waitable_event_pushed_` did not appear in my search because it was used in not loaded diffs due big size of them

alsora commented on rclcpp/src/rclcpp/experimental/timers_manager.cpp:
```diff
@@ -0,0 +1,311 @@
+
+  // Make sure the running flag is set to false when we exit from this function
+  // to allow restarting the timers thread.
+  running_ = false;
```
this should use `RCPPUTILS_SCOPE_EXIT`

mauropasse commented on rclcpp/src/rclcpp/executors/executor_notify_waitable.cpp:
```diff
@@ -99,22 +99,73 @@ ExecutorNotifyWaitable::take_data()
+  // The second argument of the callback should identify which guard condition
+  // triggered the event.
+  // We could indicate which of the guard conditions was triggered, but the executor
+  // is already going to check that.
```
Actually the executor doesn't check which guard condition was triggered, it just knows the waitable to whom it belongs.

If the executor knows which `gc` was triggered, it could more efficiently do work on the node or entity having this `gc`.
For example right know, a single `gc` triggered due addition of an entity to a node, will result on re-setting the callbacks to all nodes entities registered in the executor. We could use this info to get only the node having this `gc` and setting the callbacks to it. This is just a peformance improvement, but we already have the `notify_waitable_event_pushed_` which is mitigating this issue. 
Also if we manage to add apis to not notify the executor for every entity added, until manually done later, this will be less a problem

alsora replied:
Few comments here:
 1. Yes, the executor doesn't check which guard condition triggered, but we could easily do that (maybe I should change the `should` into `could` in the comment)
 2. This new implementation does not re-set the callbacks to all the nodes. It performs a loop on all the entities to understand which ones have been added/removed and then it does only the necessary work. By taking into account the possible improvement mentioned in (1) we could speed up the loop (i.e. loop only on one node, rather than on all nodes), but the set/unset would remain equivalent.

IMO, the fact that we don't reset everything, plus the presence of the `notify_waitable_event_pushed_` flag, should already improve the resource usage during startup.

Controlling the notification to the executor would still be important to reduce the amount of time we loop over the entities.

mjcarroll replied:
> Also if we manage to add apis to not notify the executor for every entity added, until manually done later, this will be less a problem

While we can't get this for Iron, I would like to get it in sooner rather than later in rolling while we all still have context.

alsora commented on rclcpp/src/rclcpp/experimental/executors/events_executor/events_executor.cpp:
```diff
@@ -0,0 +1,487 @@
+      }
+
+      ExecutorEvent event =
+      {notify_waitable_entity_id, gen_entity_id, ExecutorEventType::WAITABLE_EVENT, 1};
```
the linters complain if I do that

alsora commented:
Rebased with latest "exectuor_structures" branch;
 - fixed uncrustify [a642e4e](https://github.com/ros2/rclcpp/pull/2155/commits/a642e4e0fd5cdf77728abf7678db33dfe6b8590e)
 - fixed some small issues if not using events-executor [be41ece](https://github.com/ros2/rclcpp/pull/2155/commits/be41eceef3c590e4cf9cd0d4ed90ea8ba2d23cbc)


* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=18556)](http://ci.ros2.org/job/ci_linux/18556/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=13102)](http://ci.ros2.org/job/ci_linux-aarch64/13102/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=19293)](http://ci.ros2.org/job/ci_windows/19293/)

alsora commented on rclcpp/include/rclcpp/executors/executor_notify_waitable.hpp:
```diff
@@ -92,17 +92,29 @@ class ExecutorNotifyWaitable : public rclcpp::Waitable
   /**
    * \param[in] guard_condition The guard condition to add.
    */
+  RCLCPP_PUBLIC
```
Done in [a642e4e](https://github.com/ros2/rclcpp/pull/2155/commits/a642e4e0fd5cdf77728abf7678db33dfe6b8590e)

alsora commented on rclcpp/include/rclcpp/executors/executor_notify_waitable.hpp:
```diff
@@ -92,17 +92,29 @@ class ExecutorNotifyWaitable : public rclcpp::Waitable
   void
   add_guard_condition(rclcpp::GuardCondition::WeakPtr guard_condition);
 
+  RCLCPP_PUBLIC
```
Done in [a642e4e](https://github.com/ros2/rclcpp/pull/2155/commits/a642e4e0fd5cdf77728abf7678db33dfe6b8590e)

wjwwood reviewed (approved):
lgtm

clalancette reviewed (approved):
This looks good to me.  I'm going to run CI on this again since there have been host of changes since the last one.  I'll also run Windows Debug and RHEL.

clalancette commented:
All right, everything is green here so going ahead and merging.  Thanks everyone for the reviews and work.

alsora commented:
We don't currently have plans to do it.
As a reminder, there's this repository with the events-executor code that is compatible with Humble https://github.com/irobot-ros/events-executor
(although I don't remember if Humble contains all the needed features or if you still would need to build it from sources and apply some patches).

The implementation of the events-executor that got merged in this PR is slightly different from what is in the old repo.
In particular, it uses new entity collection classes that just got developed for ROS 2 Iron.

If all the prerequisites are there already, it should be possible to backport without touching existing files, and I would be more than happy to review the PRs.

russkel commented:
Would it be possible to get this into iron? Or is that something not done with non-LTS releases? Seems unfortunate to have to wait till next ROS2 release for this.

clalancette commented:
> Would it be possible to get this into iron? Or is that something not done with non-LTS releases? Seems unfortunate to have to wait till next ROS2 release for this.

This is currently in Iron.

russkel commented:
> > Would it be possible to get this into iron? Or is that something not done with non-LTS releases? Seems unfortunate to have to wait till next ROS2 release for this.
> 
> This is currently in Iron.

Ah great. I must have missed it in the release notes.

tonynajjar commented:
I think it's not in the release notes, at least it wasn't 5 days ago when I looked. It would be useful to add!

clalancette commented:
@mjcarroll is adding it in https://github.com/ros2/ros2_documentation/pull/3662
